### ISSUE TITLE
feature: Keep track if ffmpeg is installed

### ISSUE THREAD

**Issue by tygore587** [lrstanley/go-ytdlp#33]
### ✨ Describe the feature you'd like

At the moment it is checked if yt-dlp is installed and otherwise it will install it. The same would be cool to have for ffmpeg as its e.g. used to extract audio of a download or put together audio and video after downloading the best quality.

### 🌧 Is your feature request related to a problem?

I have yt-dlp configured to extract audio of a youtube video and ffmpeg is not installed, it will fail and return with the error:
```
ERROR: Postprocessing: ffprobe and ffmpeg not found. Please install or provide the path using --ffmpeg-location
exit status 1
```

So it would be a good feature to also install ffmpeg if it's not present when calling `MustInstall` or `Install`.

### 🔎 Describe alternatives you've considered

The only alternative is to give the user an error if ffmpeg is not installed if they want to run the tool or if I want to integrate it somewhere else. Or I need to make sure, I'll bring ffmpeg, so it will be always present.

### ⚠ If implemented, do you think this feature will be a breaking change to users?

No

### 🤝 Requirements

- [X] I have confirmed that someone else has not [submitted a similar feature request](../labels/enhancement).
- [X] If implemented, I believe this feature will help others, in addition to solving my problems.
- [X] I have looked into alternative solutions to the best of my ability.
- [ ] (optional) I would be willing to contribute to testing this feature if implemented, or making a PR to implement this functionality.

**Comment by lrstanley** [comment]
I've briefly looked into the feasibility of including ffmpeg, and it's not looking great so far.
1. No single download provider provides ffmpeg for all operating systems.
   - BtbN only provides git-based with a timestamp, and only includes windows/linux
   - gyan.dev only includes windows
   - johnvansickle.com provides linux static builds, however, the latest version is under a non-versioned URL, and old-releases doesn't provide the newest release, making it possible to get the latest release in a stable fashion. We'd either have to always fetch the latest, or always fetch an old version.
   - evermeet.cx provides macos based ffmpeg downloads

Given the different sources, this means a different download & extraction (including compression algorithm) solution for each, potentially different versioning for each, etc. This also doesn't include that we may not be able to properly validate all releases, as it looks like some of them aren't signed.

If I could find a single source that provides variants for all major operating systems, using all-git or all-release based builds, with some kind of signing key, that would be super helpful. Otherwise, I think this will be very low priority on the list.

**Comment by tygore587** [comment]
Thanks for your feedback :) I will also google a little bit if I find something. I also thought about it again and my idea was maybe just an option for MustInstall to check if ffmpeg is installed? So it's just returning an error and informing the user to install ffmpeg instead of failing in middle of a download process like it did for me? This would be good to know if you download an hour long (or longer) video.

**Comment by lrstanley** [comment]
I don't know if that's something that should be enabled by default (as I think there are various things you can do with yt-dlp that don't require ffmpeg), but I could see an option to install being "CheckFFMpeg" or similar.

**Comment by tygore587** [comment]
Yeah I also woudln't activate it by default (also for backwards compatibility and like you said, if someone doesn't need ffmpeg) but as an option if you want to check for it.

**Comment by tygore587** [comment]
Hm outputting a warning if somebody runs it on macOS could be enough here?

**Comment by lrstanley** [comment]
> Hm outputting a warning if somebody runs it on macOS could be maybe enough here?

I think I will make it similar to `Install()` and `MustInstall()`, so `Install()` would return an error if there is no matching os+arch combo, then the consumer of the library can decide.

I'm also tempted to make it a sub-package (with separate go.mod), as I could see someone else wanting this capability not just for go-ytdlp, to easily install ffmpeg tools.

**Comment by tygore587** [comment]
That's also a good idea in my opinion :) I think some people are using ffmpeg standalone for some other stuff/scripts and would like it to have an easy way to make sure ffmpeg is installed.

**Comment by lrstanley** [comment]
This has been implemented, though it's not as clean as I'd like, due to some of the issues I outlined above. This PR outlines the new functionality:

- https://github.com/lrstanley/go-ytdlp/pull/74
